fix(core): allow child services to inherit their @Injectable decorator

petebacondarwin opened the pull request:
fix(ivy): support sub-class services that only inherit `@Injectable`  …

 In View engine it is possible to instantiate a service that that has no
`@Injectable` decorator as long as it satisfies one of:

1) It has no dependencies and so a constructor with no parameters.
This is already supported in Ivy.
2) It has no constructor of its own and sub-classes a service which has
dependencies but has its own `@Injectable` decorator. This second
scenario was broken in Ivy.

In Ivy, previous to this commit, if a class to be instantiated did not have
its own `@Injectable` decorator and did not provide a constructor of
its own, then it would be created using `new` with no arguments -
i.e. falling back to the first scenario.

After this commit Ivy correctly uses the `ngInjectableDef` inherited
from the super-class to provide the `factory` for instantiating the
sub-class.

FW-1314

kara commented on packages/core/src/view/ng_module.ts:
```diff
@@ -107,7 +107,8 @@ export function resolveNgModuleDep(
     } else if (
-        (injectableDef = getInjectableDef(depDef.token)) && targetsModule(data, injectableDef)) {
+        (injectableDef = getInjectableDef(depDef.token, true)) &&
+        targetsModule(data, injectableDef)) {
```
It looks like this change is in ViewEngine code. We would want to make the change in `r3_injector`:

https://github.com/angular/angular/blob/master/packages/core/src/di/r3_injector.ts#L381

kara commented on packages/compiler-cli/integrationtest/bazel/injectable_def/app/test/app_spec.ts:
```diff
@@ -143,21 +143,39 @@ describe('ngInjectableDef Bazel Integration', () => {
+        expect(TestBed.get(ChildService).value).toBe(true);
+      });
+
+  fit('inherits a parent ngInjectableDef if the child class has no ngInjectableDef property of its own', () => {
```
I think we want an acceptance test in `provider_spec` for this that doesn't require `providedInRoot`.

petebacondarwin committed:
fix(ivy): support sub-class services that only inherit `@Injectable`

In View engine it is possible to instantiate a service that that has no
`@Injectable` decorator as long as it satisfies one of:

1) It has no dependencies and so a constructor with no parameters.
This is already supported in Ivy.
2) It has no constructor of its own and sub-classes a service which has
dependencies but has its own `@Injectable` decorator. This second
scenario was broken in Ivy.

In Ivy, previous to this commit, if a class to be instantiated did not have
its own `@Injectable` decorator and did not provide a constructor of
its own, then it would be created using `new` with no arguments -
i.e. falling back to the first scenario.

After this commit Ivy correctly uses the `ngInjectableDef` inherited
from the super-class to provide the `factory` for instantiating the
sub-class.

FW-1314

petebacondarwin commented on packages/core/src/di/r3_injector.ts:
```diff
@@ -378,25 +378,40 @@ export class R3Injector {
+      ngDevMode &&
+          console.warn(
+              `DEPRECATED: DI is instantiating a token "${token.name}" that inherits its @Injectable decorator but does not provide one itself.\n` +
+              'This will become an error in v9. Please add @Injectable() to the class.');
```
Tempted to move this log message into the `getInheritedInjectableDef()` function?

kara commented on packages/core/src/di/interface/defs.ts:
```diff
@@ -175,12 +175,32 @@ export function ɵɵdefineInjector(options: {factory: () => any, providers?: any
+}
+
+/**
+ * Read the `ngInjectableDef` for`type` or its ancestors.
```
```suggestion
 * Read the `ngInjectableDef` for `type` or read the `ngInjectableDef` from one of its ancestors.
```

kara commented on packages/core/src/di/interface/defs.ts:
```diff
@@ -175,12 +175,32 @@ export function ɵɵdefineInjector(options: {factory: () => any, providers?: any
+ *
+ * @param type A type which may have `ngInjectableDef`, via inheritance.
+ *
+ * @deprecated Will be removed in v9.
```
I think we decided v10, no? 

```suggestion
 * @deprecated Will be removed in v10.
```

petebacondarwin replied:
I thought that if we could deprecate it in v8 then we could kill it in v9? But happy to go with v10 for safety.

kara replied:
If the intent is to reduce dev pain, I think v10 is better because it will give people more time to migrate. Otherwise, it will break them in v9 with other Ivy changes

kara commented on packages/core/src/di/interface/defs.ts:
```diff
@@ -175,12 +175,32 @@ export function ɵɵdefineInjector(options: {factory: () => any, providers?: any
+    ngDevMode &&
+        console.warn(
+            `DEPRECATED: DI is instantiating a token "${type.name}" that inherits its @Injectable decorator but does not provide one itself.\n` +
+            'This will become an error in v9. Please add @Injectable() to the class.');
```
```suggestion
            'This will become an error in v10. Please add @Injectable() to the subclass.');
```

kara reviewed (approved):
LGTM, with documentation changes

kara commented:
[presubmit](http://test/OCL:248237231:BASE:248237247:1557877548890:aaba50fe)

kara commented on packages/core/src/di/interface/defs.ts:
```diff
@@ -175,12 +175,32 @@ export function ɵɵdefineInjector(options: {factory: () => any, providers?: any
+ */
+export function getInheritedInjectableDef<T>(type: any): ɵɵInjectableDef<T>|null {
+  if (type && type[NG_INJECTABLE_DEF]) {
+    ngDevMode &&
```
We have a closure config bug in G3 that prevents us from using ngDevMode until it can be fixed (@vikerman is working on it). For now, can you comment out `ngDevMode` and link to the Jira:

```
// TODO(FW-1307): Restore ngDevMode when G3 supports it
```

petebacondarwin replied:
So do you mean to remove the warning altogether or to log the warning even when in production mode?

kara replied:
Log the warning even in prod mode. We should be able to fix in a few days when the g3 issue is resolved.

petebacondarwin commented:
@kara - this might need another presubmit now that I have removed the `ngDevMode` and slightly refactored the code again :-)

kara reviewed (approved):
LGTM, but symbol test needs to be updated to get CI green

kara commented:
[presubmit](http://test/OCL:248237231:BASE:248378464:1557947157088:71bbe5ae)
